newhelm/ephemeral_secrets.py

```diff
@@ -0,0 +1,75 @@
+from typing import List, Mapping, Optional
+
+from pydantic import BaseModel
+
+
+RawSecrets = Mapping[str, Mapping[str, str]]
+
+
+class EphemeralSecrets:
+    def __init__(self, values: RawSecrets):
+        self._values: Optional[RawSecrets] = values
+        self.requested: List[RequestedSecret] = []
+
+    def get_required(self, scope: str, key: str, instructions: str) -> str:
+        assert self._values is not None, "Must call get_required in __init__."
+        self.requested.append(
+            RequestedSecret(
+                scope=scope,
+                key=key,
+                required=True,
+                value_present=True,
+                instructions=instructions,
+            )
+        )
+        try:
+            return self._values[scope][key]
+        except KeyError:
+            raise MissingSecretValue(scope, key, instructions)
+
+    def get_optional(self, scope: str, key: str, instructions: str) -> Optional[str]:
+        assert self._values is not None, "Must call get_optional in __init__."
+        try:
+            value = self._values[scope][key]
+        except KeyError:
+            value = None
+        self.requested.append(
+            RequestedSecret(
+                scope=scope,
+                key=key,
+                required=False,
+                value_present=value is not None,
+                instructions=instructions,
+            )
+        )
+        return value
+
+    def invalidate(self):
+        self._values = None
+
+
+class MissingSecretValue(AssertionError):
```

Comment:
This should probably subclass from `LookupError` or just `Exception`. `AssertionError` should only be used for `assert`, and `assert` should only be used for enforcing internal invariants, rather than checking user input.

Reply:
Done.